client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -139,81 +139,83 @@ private final class PoolManager[F[_], A <: Connection[F]](
-
-              case None if curTotal == maxTotal =>
-                val keys = idleQueues.keys
-                if (keys.nonEmpty) {
-                  logger.debug(
-                    s"No connections available for the desired key. Evicting random and creating a new connection: $stats")
-                  val randKey = keys.iterator.drop(Random.nextInt(keys.size)).next()
-                  getConnectionFromQueue(randKey)
-                    .fold(logger.warn(s"No connection to evict from the idleQueue for $randKey"))(
-                      _.shutdown())
-                  decrConnection(randKey)
-                  createConnection(key, callback)
-                } else {
-                  logger.debug(
-                    s"No connections available for the desired key. Adding to waitQueue: $stats")
-                  addToWaitQueue(key, callback)
-                }
-
-              case None => // we're full up. Add to waiting queue.
-                logger.debug(s"No connections available.  Waiting on new connection: $stats")
-                addToWaitQueue(key, callback)
-            }
-
-          go()
-        } else {
-          callback(Left(new IllegalStateException("Connection pool is closed")))
-        }
-      }
+            case None if numConnectionsCheckHolds(key) =>
+              F.delay(logger.debug(s"Active connection not found. Creating new one. $stats")) *>
+                F.asyncF(createConnection(key, _) *> semaphore.release)
+
+            case None if maxConnectionsPerRequestKey(key) <= 0 =>
+              semaphore.release *>
+                F.raiseError(NoConnectionAllowedException(key))
+
+            case None if curTotal == maxTotal =>
+              val keys = idleQueues.keys
+              if (keys.nonEmpty) {
+                F.delay(logger.debug(
+                  s"No connections available for the desired key. Evicting random and creating a new connection: $stats")) *>
+                  F.delay(keys.iterator.drop(Random.nextInt(keys.size)).next()).flatMap { randKey =>
+                    getConnectionFromQueue(randKey).map(_.fold(logger.warn(
+                      s"No connection to evict from the idleQueue for $randKey"))(_.shutdown())) *>
+                      decrConnection(randKey)
+                  } *>
+                  F.asyncF(createConnection(key, _) *> semaphore.release)
+
+              } else {
+                F.delay(logger.debug(
+                  s"No connections available for the desired key. Adding to waitQueue: $stats")) *>
+                  F.asyncF(addToWaitQueue(key, _) *> semaphore.release)
+              }
+
+            case None => // we're full up. Add to waiting queue.
+              F.delay(logger.debug(s"No connections available.  Waiting on new connection: $stats")) *>
+                F.asyncF(addToWaitQueue(key, _) *> semaphore.release)
+          }
+
+      F.delay(logger.debug(s"Requesting connection: $stats")) *>
+        go()
```

Comment:
@rossabaker I originally did this, but I found I couldn't match the previous behavior with that. If we use the `semaphore.withPermit` here then we'll be waiting for the `F.asyncF` calls to finish, which may take a while - and this would be a change in behavior from what the `PoolManager` did before this PR. 

[This](https://github.com/http4s/http4s/blob/master/client/src/main/scala/org/http4s/client/PoolManager.scala#L160) executes async, which immediately finishes the `synchronized` block of code. 

I think this probably could be done with `F.asyncF` surrounding the whole thing and then hitting the callback in each `case`. However, I thought that the explicit `.release` we're doing here allows a little more clarity and preciseness than the alternative.

Please let me know what your thoughts are on that. Let me know if I'm missing something 😄

Reply:
if the equivalent `getConnection` is also async (with `start`), shouldn't be enough to not have `asyncF` in that case, and use `withPermit`?